src/main/scala/rifs/business/tables/ApplicationTables.scala

```diff
@@ -74,34 +75,43 @@ class ApplicationTables @Inject()(dbConfigProvider: DatabaseConfigProvider)(impl
   }
 
+  override def fetchAppWithSection(id: ApplicationId, sectionNumber: Int): Future[Option[(ApplicationRow, Option[ApplicationSectionRow])]] = db.run {
+    appWithSectionC(id, sectionNumber).result.headOption
+  }
+
   override def fetchSection(id: ApplicationId, sectionNumber: Int): Future[Option[ApplicationSectionRow]] = db.run {
     appSectionC(id, sectionNumber).result.headOption
   }
 
   override def fetchSections(id: ApplicationId): Future[Set[ApplicationSectionRow]] = db.run(appSectionsC(id).result).map(_.toSet)
 
   override def saveSection(id: ApplicationId, sectionNumber: Int, answers: JsObject, completedAt: Option[LocalDateTime] = None): Future[Int] = {
-
-    fetchSection(id, sectionNumber).flatMap {
-      case Some(row) =>
-        preCheckSaveAnswers(row.answers, answers, completedAt) match {
-          case true => db.run(appSectionC(id, sectionNumber).update(row.copy(answers = answers, completedAt = completedAt)))
-          case false => Future.successful(1)
-        }
-      case None =>
-        db.run(applicationSectionTable += ApplicationSectionRow(None, id, sectionNumber, answers, completedAt))
+    fetchAppWithSection(id, sectionNumber).flatMap {
+      case Some((app, Some(section))) => areDifferent(section.answers, answers) || completedAt.isDefined match {
+        case true => db.run(appSectionC(id, sectionNumber).update(section.copy(answers = answers, completedAt = completedAt)))
+        case false => Future.successful(1)
+      }
+      case Some((app, None)) => db.run(applicationSectionTable += ApplicationSectionRow(None, id, sectionNumber, answers, completedAt))
+      case None => Future.successful(0)
     }
   }
 
-  override def deleteSection(id: ApplicationId, sectionNumber: Int): Future[Int] = db.run {
-    appSectionC(id, sectionNumber).delete
+  def areDifferent(obj1: JsObject, obj2: JsObject): Boolean = {
+    val flat1 = JsonHelpers.flatten("", obj1).filter(_._2.trim != "")
+    val flat2 = JsonHelpers.flatten("", obj2).filter(_._2.trim != "")
+    flat1 != flat2
   }
 
-  def preCheckSaveAnswers(obj1: JsObject, obj2: JsObject, completedAt: Option[LocalDateTime]): Boolean = {
+  def joinedAppWithSection(id: Rep[ApplicationId], sectionNumber: Rep[Int]) = for {
+    as <- applicationTable joinLeft applicationSectionTable on ((a, s) => a.id === s.applicationId && s.sectionNumber === sectionNumber) if as._1.id === id
+  } yield as
 
-    !(JsonHelpers.flatten("", obj1).filter(_._2.isEmpty == false).toList.sortBy(_._2)
-      .equals(JsonHelpers.flatten("", obj2).filter(_._2.isEmpty == false).toList.sortBy(_._2)) &&
-      completedAt.isEmpty)
+  def appWithSectionQ(id: Rep[ApplicationId], sectionNumber: Rep[Int]) = joinedAppWithSection(id, sectionNumber)
+
+  lazy val appWithSectionC = Compiled(appWithSectionQ _)
+
+  override def deleteSection(id: ApplicationId, sectionNumber: Int): Future[Int] = db.run {
```

Comment:
The calling code doesn't check if anything was actually deleted. Are we going to handle or ignore errors?